cub/test/catch2_segmented_sort_helper.cuh

```diff
@@ -0,0 +1,1421 @@
+}
+
+// Uses fully random key/value pairs. Slower than test_segments_derived.
+template <typename KeyT, typename ValueT = cub::NullType>
+void test_segments_random(
+  c2h::seed_t seed, //
+  int num_items,
+  int num_segments,
+  const int* d_begin_offsets,
+  const int* d_end_offsets)
+{
+  constexpr bool sort_pairs = !::cuda::std::is_same<ValueT, cub::NullType>::value;
+
+  CAPTURE(c2h::type_name<KeyT>(), //
+          c2h::type_name<ValueT>(),
+          sort_pairs,
+          num_items,
+          num_segments);
+
+  C2H_TIME_SECTION_INIT();
+
+  c2h::device_vector<KeyT> keys_input(num_items);
+  c2h::device_vector<KeyT> keys_output(num_items);
+
+  c2h::device_vector<ValueT> values_input;
+  c2h::device_vector<ValueT> values_output;
+  CUB_IF_CONSTEXPR(sort_pairs)
+  {
+    values_input.resize(num_items);
+    values_output.resize(num_items);
+  }
+
+  C2H_TIME_SECTION("Allocate device memory");
+
+  const bool sort_descending = GENERATE(ascending, descending);
+
+  generate_random_unsorted_inputs(seed, keys_input, values_input);
+
+  // Initialize the output values to the inputs so unused segments will be filled with the expected random values:
+  keys_output   = keys_input;
+  values_output = values_input;
+
+  C2H_TIME_SECTION_RESET();
+
+  // Since we only have offset pointers, allocate offsets first and then copy -- using the iterator constructor for
+  // cross-system copies would do a separate D->H transfer for each element:
+  c2h::host_vector<int> h_begin_offsets(num_segments);
+  thrust::copy(thrust::device_pointer_cast(d_begin_offsets),
+               thrust::device_pointer_cast(d_begin_offsets + num_segments),
+               h_begin_offsets.begin());
+  c2h::host_vector<int> h_end_offsets(num_segments);
+  thrust::copy(thrust::device_pointer_cast(d_end_offsets),
+               thrust::device_pointer_cast(d_end_offsets + num_segments),
+               h_end_offsets.begin());
+
+  // Copying a vector D->H will do a bulk copy:
+  c2h::host_vector<KeyT> h_keys_ref     = keys_input;
+  c2h::host_vector<ValueT> h_values_ref = values_input;
+
+  C2H_TIME_SECTION("D->H input arrays");
+
+  c2h::device_vector<KeyT> keys_orig     = keys_input;
+  c2h::device_vector<ValueT> values_orig = values_input;
+
+  C2H_TIME_SECTION("Clone input arrays on device");
+
+  host_sort_random_inputs(
+    sort_descending,
+    num_segments,
+    thrust::raw_pointer_cast(h_begin_offsets.data()),
+    thrust::raw_pointer_cast(h_end_offsets.data()),
+    h_keys_ref,
+    h_values_ref);
+
+  C2H_TIME_SECTION_RESET();
+
+  c2h::device_vector<KeyT> keys_ref     = h_keys_ref;
+  c2h::device_vector<ValueT> values_ref = h_values_ref;
+
+  C2H_TIME_SECTION("H->D reference arrays");
+
+  bool need_reset = false;
+
+  // Don't use GENERATE for these. We can reuse the expensive reference arrays for them:
+  for (bool stable_sort : {unstable, stable})
+  {
+    for (bool sort_buffers : {pointers, double_buffer})
```

Comment:
suggestion: as part of the Catch2 transition, I'd like to get away from this kind of manual dispatch. The motivation is the following: 
1. This code is invoked for a few type combinations and input patterns. Stable sort, for instance, is no different from unstable one at the moment. For buffers / pointers API alternatives, the code path difference is also trivial. I'd like to keep a wide test (covering more than one type combination) for double buffers / stable, and one type / a few inputs for everything else. 
2. Tests of district functionality should be in distinct test suits, so that there's no way for one test to affect state of another test.

Reply:
I agree in general with these goals, but would prefer to keep this particular test as-is.

These are combined into the same test case to reduce execution time. The time needed to do the device sort + validation is trivial, while the time taken to sort the host reference is expensive. Combining the test cases like this allows the expensive reference solution to be reused for multiple test cases. We can test the multiple variants very cheaply. Taking the slowest test that passes through this function:

```
0.000686 s: generate_random_offsets
0.003239 s: Allocate device memory
0.000466 s: generate_random_unsorted_inputs
0.007262 s: D->H input arrays
0.002009 s: Clone input arrays on device
0.514597 s: host_sort_random_inputs
0.006783 s: H->D reference arrays
0.004699 s: cub::DeviceSegmentedSort
0.001485 s: validate_sorted_random_outputs
0.000456 s: Reset input/output device arrays
0.002804 s: cub::DeviceSegmentedSort
0.000972 s: validate_sorted_random_outputs
0.000453 s: Reset input/output device arrays
0.004621 s: cub::DeviceSegmentedSort
0.001034 s: validate_sorted_random_outputs
0.000417 s: Reset input/output device arrays
0.002716 s: cub::DeviceSegmentedSort
0.001026 s: validate_sorted_random_outputs
0.557 s: "DeviceSegmentedSortPairs: Randomly sized segments, random keys/values"(0) - types_161 - 2
```

Preparing the input and reference arrays takes 535ms, while running the device algorithm and verifying the outputs 4 times with different API options takes 20ms total, roughly 1% of the total reference sort time each -- they're practically free :)

Reply:
Also, this is only for the tests with true random inputs. The tests with derived inputs execute each device sort as a separate test case.